crates/bevy_math/src/direction.rs

```diff
@@ -1090,4 +1279,26 @@ mod tests {
         );
         assert!(dir_b.is_normalized(), "Renormalisation did not work.");
     }
+
+    #[test]
+    fn dir4_creation() {
+        assert_eq!(Dir4::new(Vec4::X * 12.5), Ok(Dir4::X));
+        assert_eq!(
+            Dir4::new(Vec4::new(0.0, 0.0, 0.0, 0.0)),
+            Err(InvalidDirectionError::Zero)
+        );
+        assert_eq!(
+            Dir4::new(Vec4::new(f32::INFINITY, 0.0, 0.0, 0.0)),
+            Err(InvalidDirectionError::Infinite)
+        );
+        assert_eq!(
+            Dir4::new(Vec4::new(f32::NEG_INFINITY, 0.0, 0.0, 0.0)),
+            Err(InvalidDirectionError::Infinite)
+        );
+        assert_eq!(
+            Dir4::new(Vec4::new(f32::NAN, 0.0, 0.0, 0.0)),
+            Err(InvalidDirectionError::NaN)
+        );
+        assert_eq!(Dir4::new_and_length(Vec4::X * 6.5), Ok((Dir4::X, 6.5)));
+    }
```

Comment:
To migrate the `dir3_renorm` test I think it should be okay just to use a `Mat4` in place of the quaternion, something like:

```suggestion
    }

    #[test]
    fn dir4_renorm() {
        // Evil denormalized matrix
        let mat4 = bevy_math::Mat4::from_quat(Quat::from_euler(glam::EulerRot::XYZ, 1.0, 2.0, 3.0))
            * (1.0 + 1e-5);
        let mut dir_a = Dir4::X;
        let mut dir_b = Dir4::X;

        // We test that renormalizing an already normalized dir doesn't do anything
        assert_relative_eq!(dir_b, dir_b.fast_renormalize(), epsilon = 0.000001);

        for _ in 0..50 {
            dir_a = Dir4(mat4 * *dir_a);
            dir_b = Dir4(mat4 * *dir_b);
            dir_b = dir_b.fast_renormalize();
        }

        // `dir_a` should've gotten denormalized, meanwhile `dir_b` should stay normalized.
        assert!(
            !dir_a.is_normalized(),
            "Denormalization doesn't work, test is faulty"
        );
        assert!(dir_b.is_normalized(), "Renormalisation did not work.");
    }
```

Reply:
Thanks, I added the test to the code and it works.